### ISSUE TITLE
[analyzer] `const_constructor_param_type_mismatch` shouldn't report for for non-constant values

### ISSUE THREAD

**Issue by asashour** [dart-lang/sdk#50940]
```dart
class C {
  const C(String s);
}
void f(String s) {
  const C(s);
}
```

reports [const_constructor_param_type_mismatch](https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=const_constructor_param_type_mismatch#const_constructor_param_type_mismatch) (with a message of `A value of type 'Null' ...`) and [const_with_non_constant_argument](https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=const_with_non_constant_argument#const_with_non_constant_argument)

I think only `const_with_non_constant_argument` should be reported.

<hr>

```dart
class C {
  const C(String s);
}
void f(int i) {
  const C('$i');
}
```
reports [const_constructor_param_type_mismatch](https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=const_constructor_param_type_mismatch#const_constructor_param_type_mismatch) (with a message of `A value of type 'Null' ...`) and `INVALID_CONST`, which is commented for removal

https://github.com/dart-lang/sdk/blob/0a05e3f9994d58f3c8c83bc5721cc3bdf6ec5779/pkg/analyzer/lib/src/error/codes.g.dart#L2329-L2334

<hr>

The reason of why `const_constructor_param_type_mismatch` has a `Null` type message is that the argument value is evaluated by the `constantVisitor`

https://github.com/dart-lang/sdk/blob/0a05e3f9994d58f3c8c83bc5721cc3bdf6ec5779/pkg/analyzer/lib/src/dart/constant/evaluation.dart#L2696

which returns `ConstantEvaluationEngine._nullObject`, if the value is not constant.

In other words, whenever the constant value of the expression is null, the diagnostic shouldn't report, since its purpose seem to be when the value is only constant but its type is not matching.

If the value is not constant, then other diagnostics are triggered that non constant value is used as a constant.

This was detected in [sdk/276941](https://dart-review.googlesource.com/c/sdk/+/276941)

**Comment by bwilkerson** [comment]
> ... which returns `ConstantEvaluationEngine._nullObject`, if the value is not constant.

@scheglov 

It sounds to me like the constant evaluation engine should report when a non-constant value is found, but should then instead return a valid object of unknown value as a recovery technique. What do you think?

> ... whenever the constant value of the expression is null, the diagnostic shouldn't report ...

It's perfectly possible for the actual value of the expression to be `null`, and those cases need to be reported. The problem is that we currently can't distinguish between a real `null` being computed and a fake value of `null` being returned as a sentinel value. That's why I think we need to return something different as the sentinel value.

**Comment by scheglov** [comment]
> > ... which returns `ConstantEvaluationEngine._nullObject`, if the value is not constant.
> 
> @scheglov
> 
> It sounds to me like the constant evaluation engine should report when a non-constant value is found, but should then instead return a valid object of unknown value as a recovery technique. What do you think?

Sounds good to me.


> > ... whenever the constant value of the expression is null, the diagnostic shouldn't report ...
> 
> It's perfectly possible for the actual value of the expression to be `null`, and those cases need to be reported. The problem is that we currently can't distinguish between a real `null` being computed and a fake value of `null` being returned as a sentinel value. That's why I think we need to return something different as the sentinel value.

I agree.

**Comment by asashour** [comment]
[sdk/279397](https://dart-review.googlesource.com/c/sdk/+/279397)

If a non-constant value is provided, then the hint isn't reported.

If a constant value evaluates to `null`, then it is reported.

If an expected argument is not specified, it is not reported, and other hints (e.g. `NOT_ENOUGH_POSITIONAL_ARGUMENTS_NAME_SINGULAR` are reported in this case).

Interestingly, one test case without null safety, now correctly triggers a new hint.
